sdk/core/Azure.Core/src/Shared/BearerTokenChallengeAuthenticationPolicy.cs

```diff
@@ -46,28 +47,40 @@ internal BearerTokenChallengeAuthenticationPolicy(TokenCredential credential, IE
         }
 
         /// <inheritdoc />
-        public override ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
-            return ProcessAsync(message, pipeline, true);
+            await PreProcessAsync(message, true).ConfigureAwait(false);
+            await ProcessAsync(message, pipeline, true).ConfigureAwait(false);
         }
 
         /// <inheritdoc />
         public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
+            PreProcessAsync(message, false).EnsureCompleted();
             ProcessAsync(message, pipeline, false).EnsureCompleted();
         }
 
+        /// <summary>
+        /// Executes a pre-processing step on the <paramref name="message"/> before <see cref="ProcessAsync(HttpMessage, ReadOnlyMemory{HttpPipelinePolicy})"/> or <see cref="Process(HttpMessage, ReadOnlyMemory{HttpPipelinePolicy})"/> is called.
+        /// </summary>
+        /// <param name="message">The <see cref="HttpMessage"/> this policy would be applied to.</param>
+        /// <param name="async">Indicates whether the method was called from an asynchronous context.</param>
+        /// <returns>The <see cref="ValueTask"/> representing the asynchronous operation.</returns>
+        protected virtual Task PreProcessAsync(HttpMessage message, bool async)
```

Comment:
I almost think this should be named `AuthenticateRequestAsync` to match `AuthenticateReqeustFromChallenge` so it's clear they serve the same purpose. Perhaps the helper method could be called something more specific like `AddAuthorizationHeader`?

Reply:
Initially, my concern with that name was that it implies that the only reason for implementers to override this is to authenticate the request and I wasn't sure if that is true. However, if it is true, I think the rename makes sense and we could make the change discussed in the other comment regarding the base implementation as well.

Thinking about it a bit more, I think this change does clarify the purpose of the method and provides better guidance to implementers. Thanks for the suggestion!